EIPS/eip-4762.md

```diff
@@ -35,155 +35,266 @@ def get_storage_slot_tree_keys(storage_key: int) -> [int, int]:
+  match opcode:
+    case OP_JUMP:
+      dest = stack.peek(0)
+      # The destination of a jump is not considered to be accessed if it is beyond the code
+      if dest < len(code):
+        # The destination of a `JUMP` is considered to be accessed, even if the destination is not a jumpdest or is inside pushdata
+        add_read_access_event(dest)
+
+    case OP_JUMPI:
+      dest = stack.peek(0)
+      cond = stack.peek(1)
+      # The destination of a jump is not considered to be accessed if it is beyond the code
+      if dest < len(code):
+        dest_chunk = dest // CHUNK_SIZE
+        # The destination of a positively evaluated `JUMPI` is considered to be accessed, even if the destination is not a jumpdest or is inside pushdata
+        # The destination of a `JUMPI` is not considered to be accessed if the jump conditional is `false`
+        if op == OP_JUMP or cond:
+          add_read_access_event(contract_address, (dest_chunk + CODE_OFFSET) // 256, (dest_chunk + CODE_OFFSET) % 256)
+
+    case OP_PUSH1 | OP_PUSH2 | ... | OP_PUSH32:
+      nbytes = opcode - OP_PUSH0 # determine the number of extra bytes to touch
+      for i in range(nbytes):
+        immediate_chunk = (pc+i) // CHUNK_SIZE
+        if immediate_chunk <= (len(code)+CHUNK_SIZE-1) // CHUNK_SIZE:
+          add_read_access_event(contract_address, (immediate_chunk + CODE_OFFSET) // 256, (immediate_chunk + CODE_OFFSET) % 256)
+
+    case OP_CODECOPY | OP_EXTCODECOPY:
+      addr = contract_address
+      size = 0
+      offset = i0
+      if opcode == OP_EXTCODECOPY:
+        addr = stack.peek(0)
+        offset = stack.peek(2)
+        size = stack.peek(3)
+      else:
+        offset = stack.peek(1)
+        size = stack.peek(2)
+
+      for i in range(size):
+        copied_byte_chunk = (offset+i) // CHUNK_SIZE 
+        if copied_byte_chunk <= (len(code)+CHUNK_SIZE-1) // CHUNK_SIZE:
+          add_read_access_event(contract_address, (copied_byte_chunk + CODE_OFFSET) // 256, (copied_byte_chunk + CODE_OFFSET) % 256)
+
+    case OP_CODESIZE | OP_EXTCODESIZE | OP_EXTCODEHASH:
+      # None of these operations access any chunk, it is
+      # explicitly documented here as a no-op.
+
+  charge_gas()
 ```
-(contract_address, 0, BASIC_DATA_LEAF_KEY)
-(contract_address, 0, CODEHASH_LEAF_KEY)
-```
-
-#### Write events for storage
-
-`SSTORE` opcodes with a given `address` and `key` process a write event of the form
 
-```
-(address, tree_key, sub_key)
-```
+Note: since no access list existed for code up until this EIP, note that no warm costs are charged for code accesses.
 
-Where `tree_key` and `sub_key` are computed as `tree_key, sub_key = get_storage_slot_tree_keys(address, key)`
+When a contract is created, access chunks `0 ... (len(code)+30)//31`:
 
-#### Write events for code
+```python3
+def emit_code_write_access_events(address: Address, code: bytes):
+  last_chunk_num = (len(code) + CHUNK_SIZE - 1) // CHUNK_SIZE
```

Comment:
```suggestion
    # Calculate the number of code chunks required, rounding up for any partial final chunk.
    # This ensures that all code bytes are covered, even if the code size is not a multiple of CHUNK_SIZE.
  last_chunk_num = (len(code) + CHUNK_SIZE - 1) // CHUNK_SIZE
```
This is a caveat that is easy to miss and can lead to hard to spot bugs, imo worth adding a comment